Source/Libraries/openXDA.APIAuthentication/APIAuthenticationMiddleware.cs

```diff
@@ -134,14 +157,41 @@ public async override Task Invoke(IOwinContext context)
             IOwinRequest request = context.Request;
             AuthorizationHeader authorization = new AuthorizationHeader(request.Headers);
 
-            if (IsAuthorized(authorization))
+            if (!authorization.Valid)
+            {
+                await Next.Invoke(context);
+                return;
+            }
+
+            if (IsAuthorized(authorization) && !UseImpersonation(authorization))
             {
                 string apiKey = authorization.APIKey;
                 ISecurityProvider provider = new HostSecurityProvider(apiKey);
                 SecurityIdentity identity = new SecurityIdentity(provider);
                 SecurityPrincipal principal = new SecurityPrincipal(identity);
                 context.Request.User = principal;
             }
+            else if (IsAuthorized(authorization) && UseImpersonation(authorization))
+            {
+                string impersonatedUser = authorization.ImpersonationToken;
+                ISecurityProvider provider = new ImpersonationSecurityProvider(impersonatedUser);
+
+                if(!provider.RefreshData())
+                {
+                    await Next.Invoke(context);
+                    return;
+                }
```

Comment:
Every one of these early-return edge cases also call `Next.Invoke(context)`. We could simplify a bit if we have `Invoke()` call another method that isn't responsible for calling `Next.Invoke(context)`.

```c#
public async override Task Invoke(IOwinContext context)
{
    Authorize(context.Request);
    await Next.Invoke(context);
}

private void Authorize(IOwinRequest request)
{
    AuthorizationHeader authorization = new AuthorizationHeader(request.Headers);

    if (!authorization.Valid || !IsAuthorized(authorization))
        return;

    ...
}
```